File: app/src/main/java/com/phlox/tvwebbrowser/activity/downloads/DownloadsActivity.java
package com.phlox.tvwebbrowser.activity.downloads;

import android.app.ListActivity;
import android.content.ActivityNotFoundException;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.net.Uri;
import android.os.Bundle;
import android.os.Environment;
import android.os.IBinder;
import android.support.v4.content.FileProvider;
import android.view.Gravity;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.MimeTypeMap;
import android.widget.AbsListView;
import android.widget.AdapterView;
import android.widget.BaseAdapter;
import android.widget.FrameLayout;
import android.widget.PopupMenu;
import android.widget.TextView;

import com.phlox.asql.ASQL;
import com.phlox.tvwebbrowser.BuildConfig;
import com.phlox.tvwebbrowser.R;
import com.phlox.tvwebbrowser.model.Download;
import com.phlox.tvwebbrowser.service.downloads.DownloadService;
import com.phlox.tvwebbrowser.utils.Utils;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

import static com.phlox.tvwebbrowser.R.string.url;

public class DownloadsActivity extends ListActivity implements AdapterView.OnItemClickListener, DownloadService.Listener, AdapterView.OnItemLongClickListener {
    private TextView tvPlaceholder;
    private DownloadListAdapter adapter;
    private ASQL asql;
    private boolean loading = false;
    private DownloadService downloadsService;
    private List<DownloadService.Listener> listeners = new ArrayList<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_downloads);
        tvPlaceholder = (TextView) findViewById(R.id.tvPlaceholder);

        adapter = new DownloadListAdapter(this);
        setListAdapter(adapter);
        asql = ASQL.getDefault(this);

        getListView().setOnScrollListener(onListScrollListener);
        getListView().setOnItemClickListener(this);
        getListView().setOnItemLongClickListener(this);

        loadItems();
    }

    @Override
    protected void onStart() {
        super.onStart();
        bindService(new Intent(this, DownloadService.class), downloadsServiceConnection, Context.BIND_AUTO_CREATE);
    }

    @Override
    protected void onStop() {
        unbindService(downloadsServiceConnection);
        super.onStop();
    }

    @Override
    public void onItemClick(AdapterView<?> adapterView, View view, int i, long l) {
        DownloadListItemView v = (DownloadListItemView) view;
        if (v.download.isDateHeader) {
            return;
        }
        File file = new File(v.download.filepath);
        if (!file.exists()) {
            Utils.showToast(this, R.string.file_not_found);
        }
        if (v.download.size != v.download.bytesReceived) {
            return;
        }
        //Uri pathUri = Uri.fromFile(file);
        Uri pathUri = FileProvider.getUriForFile(DownloadsActivity.this,
                BuildConfig.APPLICATION_ID + ".provider",
                file);
        Intent openIntent = new Intent(Intent.ACTION_VIEW);
        String extension = getFileExtension(v.download.filepath);
        if (extension != null) {
            openIntent.setDataAndType(pathUri, MimeTypeMap.getSingleton().getMimeTypeFromExtension(extension));
        } else {
            openIntent.setData(pathUri);
        }
        openIntent.setFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
        try {
            startActivity(openIntent);
        } catch (ActivityNotFoundException e) {
            Utils.showToast(this, getString(R.string.no_app_for_file_type));
        }
    }

    @Override
    public boolean onItemLongClick(AdapterView<?> adapterView, View view, int i, long l) {
        final DownloadListItemView v = (DownloadListItemView) view;
        if (v.download.isDateHeader) {
            return true;
        }
        if (v.download.size == Download.BROKEN_MARK || v.download.size == Download.CANCELLED_MARK ||
                v.download.size == v.download.bytesReceived) {
            showFinishedDownloadOptionsPopup(v);
        } else {
            showUnfinishedDownloadOptionsPopup(v);
        }
        return true;
    }

    private void showUnfinishedDownloadOptionsPopup(final DownloadListItemView v) {
        PopupMenu pm = new PopupMenu(this, v, Gravity.BOTTOM);
        pm.getMenu().add(R.string.cancel);
        pm.setOnMenuItemClickListener(new PopupMenu.OnMenuItemClickListener() {
            @Override
            public boolean onMenuItemClick(MenuItem menuItem) {
                downloadsService.cancelDownload(v.download);
                return true;
            }
        });
        pm.show();
    }

    private void showFinishedDownloadOptionsPopup(final DownloadListItemView v) {
        PopupMenu pm = new PopupMenu(this, v, Gravity.BOTTOM);
        pm.getMenu().add(0, 1, 1, R.string.open_folder);
        pm.getMenu().add(0, 2, 2, R.string.delete);
        pm.setOnMenuItemClickListener(new PopupMenu.OnMenuItemClickListener() {
            @Override
            public boolean onMenuItemClick(MenuItem menuItem) {
                switch (menuItem.getItemId()) {
                    case 1:
                        openFolder(v);
                        break;
                    case 2:
                        deleteItem(v);
                        break;
                }
                return true;
            }
        });
        pm.show();
    }

    private void openFolder(DownloadListItemView v) {
        Uri uri = Uri.parse(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS).getAbsolutePath());
        Intent intent = new Intent(Intent.ACTION_VIEW);
        intent.setDataAndType(uri, "resource/folder");
        if (intent.resolveActivityInfo(getPackageManager(), 0) != null) {
            startActivity(intent);
        } else {
            Utils.showToast(this, R.string.no_file_explorer_msg);
        }
    }

    private void deleteItem(final DownloadListItemView v) {
        new File(v.download.filepath).delete();
        asql.delete(v.download, new ASQL.ResultCallback<Integer>() {
            @Override
            public void onDone(Integer result, Exception exception) {
                if (exception != null) {
                    Utils.showToast(DownloadsActivity.this, R.string.error);
                } else {
                    adapter.remove(v.download);
                }
            }
        });
    }

    @Override
    public void onDownloadUpdated(Download downloadInfo) {
        for (int i = 0; i < listeners.size(); i++) {
            listeners.get(i).onDownloadUpdated(downloadInfo);
        }
    }

    @Override
    public void onDownloadError(Download downloadInfo, int responseCode, String responseMessage) {
        for (int i = 0; i < listeners.size(); i++) {
            listeners.get(i).onDownloadError(downloadInfo, responseCode, responseMessage);
        }
    }

    @Override
    public void onAllDownloadsComplete() {
    }

    public void registerListener(DownloadService.Listener listener) {
        listeners.add(listener);
    }

    public void unregisterListener(DownloadService.Listener listener) {
        listeners.remove(listener);
    }

    private void loadItems() {
        if (loading) {
            return;
        }
        loading = true;

        asql.queryAll(Download.class, "SELECT * FROM downloads ORDER BY time DESC LIMIT 100 OFFSET ?",
                sqlCallback, Long.toString(adapter.getRealCount()));
    }

    private ASQL.ResultCallback<List<Download>> sqlCallback = new ASQL.ResultCallback<List<Download>>() {
        @Override
        public void onDone(List<Download> result, Exception error) {
            loading = false;
            if (result != null) {
                if (!result.isEmpty()) {
                    tvPlaceholder.setVisibility(View.GONE);
                    adapter.addItems(result);
                    getListView().requestFocus();
                }
            } else {
                Utils.showToast(DownloadsActivity.this, R.string.error);
            }
        }
    };

    AbsListView.OnScrollListener onListScrollListener = new AbsListView.OnScrollListener() {
        @Override
        public void onScrollStateChanged(AbsListView view, int scrollState) {

        }

        @Override
        public void onScroll(AbsListView view, int firstVisibleItem, int visibleItemCount, int totalItemCount) {
            if (totalItemCount != 0 && firstVisibleItem + visibleItemCount >= totalItemCount - 1) {
                loadItems();
            }
        }
    };

    ServiceConnection downloadsServiceConnection = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName name, IBinder service) {
            DownloadService.Binder binder = (DownloadService.Binder) service;
            downloadsService = binder.getService();
            downloadsService.registerListener(DownloadsActivity.this);
        }

        @Override
        public void onServiceDisconnected(ComponentName name) {
            downloadsService.unregisterListener(DownloadsActivity.this);
            downloadsService = null;
        }
    };

    static String getFileExtension(String filePath) {
        String result = "";
        String[] parts = filePath.split("\\.");
        if (parts.length > 0)
            result = parts[parts.length - 1];
        return result;
    }
}

File: app/src/main/java/com/phlox/tvwebbrowser/activity/downloads/DownloadsActivity.kt
package com.phlox.tvwebbrowser.activity.downloads

import android.app.Activity
import android.app.AlertDialog
import android.content.*
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.os.Environment
import android.provider.Settings
import android.util.Log
import android.view.Gravity
import android.view.View
import android.webkit.MimeTypeMap
import android.widget.AbsListView
import android.widget.AdapterView
import android.widget.PopupMenu
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.FileProvider
import androidx.lifecycle.lifecycleScope
import com.phlox.tvwebbrowser.BuildConfig
import com.phlox.tvwebbrowser.R
import com.phlox.tvwebbrowser.databinding.ActivityDownloadsBinding
import com.phlox.tvwebbrowser.model.Download
import com.phlox.tvwebbrowser.utils.Utils
import com.phlox.tvwebbrowser.utils.activemodel.ActiveModelsRepository
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.io.File
import java.util.*

class DownloadsActivity : AppCompatActivity(), AdapterView.OnItemClickListener, ActiveDownloadsModel.Listener, AdapterView.OnItemLongClickListener{
    private lateinit var vb: ActivityDownloadsBinding
    private lateinit var adapter: DownloadListAdapter
    private val listeners = ArrayList<ActiveDownloadsModel.Listener>()

    private lateinit var activeDownloadsModel: ActiveDownloadsModel
    private lateinit var downloadsHistoryModel: DownloadsHistoryModel

    internal var onListScrollListener: AbsListView.OnScrollListener = object : AbsListView.OnScrollListener {
        override fun onScrollStateChanged(view: AbsListView, scrollState: Int) {

        }

        override fun onScroll(view: AbsListView, firstVisibleItem: Int, visibleItemCount: Int, totalItemCount: Int) {
            if (totalItemCount != 0 && firstVisibleItem + visibleItemCount >= totalItemCount - 1) {
                downloadsHistoryModel.loadNextItems()
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        Log.d(TAG, "onCreate this:" + System.identityHashCode(this))
        vb = ActivityDownloadsBinding.inflate(layoutInflater)
        setContentView(vb.root)

        activeDownloadsModel = ActiveModelsRepository.get(ActiveDownloadsModel::class, this)
        downloadsHistoryModel = ActiveModelsRepository.get(DownloadsHistoryModel::class, this)

        adapter = DownloadListAdapter(this)
        vb.listView.adapter = adapter

        vb.listView.setOnScrollListener(onListScrollListener)
        vb.listView.onItemClickListener = this
        vb.listView.onItemLongClickListener = this

        downloadsHistoryModel.lastLoadedItems.subscribe(this, false, {
            if (it.isNotEmpty()) {
                vb.tvPlaceholder.visibility = View.GONE
                adapter.addItems(it)
                vb.listView.requestFocus()
            }
        })

        if (downloadsHistoryModel.allItems.isEmpty()) {
            downloadsHistoryModel.loadNextItems()
        } else {
            vb.tvPlaceholder.visibility = View.GONE
            adapter.addItems(downloadsHistoryModel.allItems)
            vb.listView.requestFocus()
        }
    }

    override fun onResume() {
        super.onResume()
        activeDownloadsModel.registerListener(this)
    }

    override fun onPause() {
        activeDownloadsModel.unregisterListener(this@DownloadsActivity)
        super.onPause()
    }

    override fun onItemClick(adapterView: AdapterView<*>, view: View, i: Int, l: Long) {
        val v = view as DownloadListItemView
        if (v.download == null) return
        if (v.download?.isDateHeader!!) {
            return
        }
        val file = File(v.download?.filepath!!)
        if (!file.exists()) {
            Utils.showToast(this, R.string.file_not_found)
        }
        if (v.download?.size != v.download?.bytesReceived) {
            return
        }
        val pathUri = FileProvider.getUriForFile(this@DownloadsActivity,
                BuildConfig.APPLICATION_ID + ".provider",
                file)
        val openIntent = Intent(Intent.ACTION_VIEW)
        val extension = getFileExtension(v.download?.filepath!!)
        if (extension != null) {
            openIntent.setDataAndType(pathUri, MimeTypeMap.getSingleton().getMimeTypeFromExtension(extension))
        } else {
            openIntent.data = pathUri
        }
        openIntent.flags = Intent.FLAG_GRANT_READ_URI_PERMISSION
        try {
            startActivity(openIntent)
        } catch (e: ActivityNotFoundException) {
            Utils.showToast(this, getString(R.string.no_app_for_file_type))
        }

    }

    override fun onItemLongClick(adapterView: AdapterView<*>, view: View, i: Int, l: Long): Boolean {
        val v = view as DownloadListItemView
        if (v.download == null) return false
        if (v.download?.isDateHeader!!) {
            return true
        }
        if (v.download?.size == Download.BROKEN_MARK || v.download?.size == Download.CANCELLED_MARK ||
                v.download?.size == v.download?.bytesReceived) {
            showFinishedDownloadOptionsPopup(v)
        } else {
            showUnfinishedDownloadOptionsPopup(v)
        }
        return true
    }

    private fun showUnfinishedDownloadOptionsPopup(v: DownloadListItemView) {
        val pm = PopupMenu(this, v, Gravity.BOTTOM)
        pm.menu.add(R.string.cancel)
        pm.setOnMenuItemClickListener {
            activeDownloadsModel.cancelDownload(v.download!!)
            true
        }
        pm.show()
    }

    private fun showFinishedDownloadOptionsPopup(v: DownloadListItemView) {
        val pm = PopupMenu(this, v, Gravity.BOTTOM)
        if (v.download?.filename!!.endsWith(".apk", true) &&
                v.download?.size == v.download?.bytesReceived) {
            pm.menu.add(0, 0, 0, R.string.install)
        }
        pm.menu.add(0, 1, 1, R.string.open_folder)
        pm.menu.add(0, 2, 2, R.string.delete)
        pm.setOnMenuItemClickListener { menuItem ->
            when (menuItem.itemId) {
                0 -> installAPK(v.download!!)
                1 -> openFolder(v)
                2 -> deleteItem(v)
            }
            true
        }
        pm.show()
    }

    private fun installAPK(download: Download) {
        val canInstallFromOtherSources = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            //packageManager.canRequestPackageInstalls()
            true
        } else
            Settings.Secure.getInt(this.contentResolver, Settings.Secure.INSTALL_NON_MARKET_APPS) == 1

        if(canInstallFromOtherSources) {
            launchInstallAPKActivity(this, download)
        } else {
            AlertDialog.Builder(this)
                    .setTitle(R.string.app_name)
                    .setMessage(if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O)
                        R.string.turn_on_unknown_sources_for_app else R.string.turn_on_unknown_sources)
                    .setPositiveButton(android.R.string.ok, DialogInterface.OnClickListener { dialog, which -> run {
                        val intentSettings = Intent()
                        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                            intentSettings.action = Settings.ACTION_MANAGE_UNKNOWN_APP_SOURCES
                            intentSettings.data = Uri.parse("package:${BuildConfig.APPLICATION_ID}")
                        } else {
                            intentSettings.action = Settings.ACTION_SECURITY_SETTINGS
                        }
                        intentSettings.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK or Intent.FLAG_ACTIVITY_NEW_TASK)
                        try {
                            startActivityForResult(intentSettings, REQUEST_CODE_UNKNOWN_APP_SOURCES)
                        } catch (e: Exception) {
                            e.printStackTrace()
                            Toast.makeText(this, R.string.error, Toast.LENGTH_SHORT).show()
                        }
                    }})
                    .show()

        }
    }

    private fun openFolder(v: DownloadListItemView) {
        val uri = Uri.parse(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS).absolutePath)
        val intent = Intent(Intent.ACTION_VIEW)
        intent.setDataAndType(uri, "resource/folder")
        if (intent.resolveActivityInfo(packageManager, 0) != null) {
            startActivity(intent)
        } else {
            Utils.showToast(this, R.string.no_file_explorer_msg)
        }
    }

    private fun deleteItem(v: DownloadListItemView) = lifecycleScope.launch(Dispatchers.Main) {
        v.download?.let {
            activeDownloadsModel.deleteItem(it)
            adapter.remove(it)
        }
    }

    override fun onDownloadUpdated(downloadInfo: Download) {
        for (i in listeners.indices) {
            listeners[i].onDownloadUpdated(downloadInfo)
        }
    }

    override fun onDownloadError(downloadInfo: Download, responseCode: Int, responseMessage: String) {
        for (i in listeners.indices) {
            listeners[i].onDownloadError(downloadInfo, responseCode, responseMessage)
        }
    }

    override fun onAllDownloadsComplete() {}

    fun registerListener(listener: ActiveDownloadsModel.Listener) {
        listeners.add(listener)
    }

    fun unregisterListener(listener: ActiveDownloadsModel.Listener) {
        listeners.remove(listener)
    }

    companion object {
        const val REQUEST_CODE_UNKNOWN_APP_SOURCES = 10007
        const val REQUEST_CODE_INSTALL_PACKAGE = 10008
        val TAG = DownloadsActivity::class.java.simpleName

        internal fun getFileExtension(filePath: String): String? {
            var result = ""
            val parts = filePath.split("\\.".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            if (parts.size > 0)
                result = parts[parts.size - 1]
            return result
        }

        fun launchInstallAPKActivity(activity: Activity, download: Download) {
            val file = File(download.filepath)
            val mimeType = MimeTypeMap.getSingleton().getMimeTypeFromExtension(file.extension)
            val apkURI = FileProvider.getUriForFile(
                    activity,
                    activity.applicationContext.packageName + ".provider", file)

            val install = Intent(Intent.ACTION_INSTALL_PACKAGE)
            install.setDataAndType(apkURI, mimeType)
            install.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
            try {
                activity.startActivityForResult(install, REQUEST_CODE_INSTALL_PACKAGE)//we are not using result for now
            } catch (e: ActivityNotFoundException) {
                Toast.makeText(activity, R.string.error, Toast.LENGTH_SHORT).show()
            }
        }
    }
}
